<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introducing R and RStudio • Metagenomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introducing R and RStudio">
<meta property="og:description" content="Metagenomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Metagenomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01-introduction.html">Introducing R and RStudio</a>
    </li>
    <li>
      <a href="../articles/02-r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="../articles/03-basics-factors-dataframes.html">R Basics continued - factors and data frames</a>
    </li>
    <li>
      <a href="../articles/04-dplyr.html">Aggregating and Analyzing Data with dplyr</a>
    </li>
    <li>
      <a href="../articles/05-data-visualization.html">Data Visualization with ggplot2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="01-introduction_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introducing R and RStudio</h1>
                        <h4 class="author">Klaus Schliep</h4>
            
            <h4 class="date">Mai 17, 2021</h4>
      
      
      <div class="hidden name"><code>01-introduction.Rmd</code></div>

    </div>

    
    
<p>This class is based on the <a href="https://datacarpentry.org/genomics-r-intro">https://datacarpentry.org/genomics-r-intro</a>. Mainly just rendered differently.</p>
<div id="getting-ready-to-use-r-for-the-first-time" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-ready-to-use-r-for-the-first-time" class="anchor"></a>Getting ready to use R for the first time</h2>
<p>In this lesson we will take you through the very first things you need to get R working.</p>
</div>
<div id="a-brief-history-of-r" class="section level2">
<h2 class="hasAnchor">
<a href="#a-brief-history-of-r" class="anchor"></a>A Brief History of R</h2>
<p><a href="https://en.wikipedia.org/wiki/R_(programming_language)">R</a> has been around since 1995, and was created by Ross Ihaka and Robert Gentleman at the University of Auckland, New Zealand. R is based off the S programming language developed at Bell Labs and was developed to teach intro statistics.</p>
</div>
<div id="advantages-of-using-r" class="section level2">
<h2 class="hasAnchor">
<a href="#advantages-of-using-r" class="anchor"></a>Advantages of using R</h2>
<p>At more than 20 years old, R is fairly mature and <a href="https://www.tiobe.com/tiobe-index/r/">growing in popularity</a>. However, programming isn’t a popularity contest. Here are key advantages of analyzing data in R:</p>
<ul>
<li>
<strong>R is <a href="https://en.wikipedia.org/wiki/Open-source_software">open source</a></strong>. This means R is free - an advantage if you are at an institution where you have to pay for your own MATLAB or SAS license. Open source, is important to your colleagues in parts of the world where expensive software in inaccessible. It also means that R is actively developed by a community (see <a href="https://www.r-project.org/">r-project.org</a>), and there are regular updates.</li>
<li>
<strong>R is widely used</strong>. Ok, maybe programming is a popularity contest. Because, R is used in many areas (not just bioinformatics), you are more likely to find help online when you need it. Chances are, almost any error message you run into, someone else has already experienced.</li>
<li>
<strong>R is powerful</strong>. R runs on multiple platforms (Windows/MacOS/Linux). It can work with much larger datasets than popular spreadsheet programs like Microsoft Excel, and because of its scripting capabilities is far more reproducible. Also, there are thousands of available software packages for science, including genomics and other areas of life science.</li>
</ul>
<blockquote>
<h2 id="discussion-your-experience">Discussion: Your experience</h2>
<p>What has motivated you to learn R? Have you had a research question for which spreadsheet programs such as Excel have proven difficult to use, or where the size of the data set created issues?</p>
</blockquote>
</div>
<div id="introducing-rstudio" class="section level2">
<h2 class="hasAnchor">
<a href="#introducing-rstudio" class="anchor"></a>Introducing RStudio</h2>
<p>In these lessons, we will be making use of a software called <a href="https://www.rstudio.com/products/RStudio/">RStudio</a>, an <a href="https://en.wikipedia.org/wiki/Integrated_development_environment">Integrated Development Environment (IDE)</a>. RStudio, like most IDEs, provides a graphical interface to R, making it more user-friendly, and providing dozens of useful features.</p>
<p>Open RStudio and the interface should look similar to this:</p>
<p><img src="fig/RStudio.png" alt="rstudio default session" style="width:1000px;"></p>
</div>
<div id="create-an-rstudio-project" class="section level2">
<h2 class="hasAnchor">
<a href="#create-an-rstudio-project" class="anchor"></a>Create an RStudio project</h2>
<p>One of the first benefits we will take advantage of in RStudio is something called an <strong>RStudio Project</strong>. An RStudio project allows you to more easily:</p>
<ul>
<li>Save data, files, variables, packages, etc. related to a specific analysis project</li>
<li>Restart work where you left off</li>
<li>Collaborate, especially if you are using version control such as <a href="http://swcarpentry.github.io/git-novice/">git</a>.</li>
</ul>
<ol style="list-style-type: decimal">
<li>To create a project, go to the <kbd>File</kbd> menu, and click <kbd>New Project…</kbd>.</li>
</ol>
<p><img src="fig/new_project_window.png" alt="rstudio default session" style="width: 600px;"></p>
<ol start="2" style="list-style-type: decimal">
<li><p>In the window that opens select <strong>New Directory</strong>, then <strong>New Project</strong>. For “Directory name:” enter <strong>dc_genomics_r</strong>. For “Create project as subdirectory of”, you may leave the default, which is your home directory “~”.</p></li>
<li><p>Finally click <kbd>Create Project</kbd>. In the “Files” tab of your output pane (more about the RStudio layout in a moment), you should see an RStudio project file, <strong>dc_genomics_r.Rproj</strong>. All RStudio projects end with the “<strong>.Rproj</strong>” file extension.</p></li>
</ol>
</div>
<div id="creating-your-first-r-script" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-your-first-r-script" class="anchor"></a>Creating your first R script</h2>
<p>Now that we are ready to start exploring R, we will want to keep a record of the commands we are using. To do this we can create an R script:</p>
<p>Click the <kbd>File</kbd> menu and select <kbd>New File</kbd> and then <kbd>R Script</kbd>. Before we go any further, save your script by clicking the save/disk icon that is in the bar above the first line in the script editor, or click the <kbd>File</kbd> menu and select <kbd>save</kbd>. In the “Save File” window that opens, name your file <strong>“genomics_r_basics”</strong>. The new script <strong>genomics_r_basics.R</strong> should appear under “files” in the output pane. By convention, R scripts end with the file extension <strong>.R</strong>.</p>
</div>
<div id="overview-and-customization-of-the-rstudio-layout" class="section level2">
<h2 class="hasAnchor">
<a href="#overview-and-customization-of-the-rstudio-layout" class="anchor"></a>Overview and customization of the RStudio layout</h2>
<p>Here are the major windows (or panes) of the RStudio environment:</p>
<p><img src="fig/rstudio_session_4pane_layout.png" alt="rstudio default session" style="width: 1000px;"></p>
<ul>
<li>
<strong>Source</strong>: This pane is where you will write/view R scripts. Some outputs (such as if you view a dataset using <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code>) will appear as a tab here.</li>
<li>
<strong>Console/Terminal</strong>: This is actually where you see the execution of commands. This is the same display you would see if you were using R at the command line without RStudio. You can work interactively (i.e. enter R commands here), but for the most part we will run a script (or lines in a script) in the source pane and watch their execution and output here. The “Terminal” tab give you access to the BASH terminal (the Linux operating system, unrelated to R).</li>
<li>
<strong>Environment/History</strong>: Here, RStudio will show you what datasets and objects (variables) you have created and which are defined in memory. You can also see some properties of objects/datasets such as their type and dimensions. The “History” tab contains a history of the R commands you’ve executed R.</li>
<li>
<strong>Files/Plots/Packages/Help</strong>: This multipurpose pane will show you the contents of directories on your computer. You can also use the “Files” tab to navigate and set the working directory. The “Plots” tab will show the output of any plots generated. In “Packages” you will see what packages are actively loaded, or you can attach installed packages. “Help” will display help files for R functions and packages.</li>
</ul>
<blockquote>
<h2 id="tip-uploads-and-downloads-in-the-cloud">Tip: Uploads and downloads in the cloud</h2>
<p>In the “Files” tab you can select a file and download it from your cloud instance (click the “more” button) to your local computer. Uploads are also possible.</p>
</blockquote>
<p>All of the panes in RStudio have configuration options. For example, you can minimize/maximize a pane, or by moving your mouse in the space between panes you can resize as needed. The most important customization options for pane layout are in the <kbd>View</kbd> menu. Other options such as font sizes, colors/themes, and more are in the <kbd>Tools</kbd> menu under <kbd>Global Options</kbd>.</p>
<blockquote>
<h2 id="you-are-working-with-r">You are working with R</h2>
<p>Although we won’t be working with R at the terminal, there are lots of reasons to. For example, once you have written an RScript, you can run it at any Linux or Windows terminal without the need to start up RStudio. We don’t want you to get confused - RStudio runs R, but R is not RStudio. For more on running an R Script at the terminal see this <a href="https://swcarpentry.github.io/r-novice-inflammation/05-cmdline/">Software Carpentry lesson</a>.</p>
</blockquote>
</div>
<div id="getting-to-work-with-r-navigating-directories" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-to-work-with-r-navigating-directories" class="anchor"></a>Getting to work with R: navigating directories</h2>
<p>Now that we have covered the more aesthetic aspects of RStudio, we can get to work using some commands. We will write, execute, and save the commands we learn in our <strong>genomics_r_basics.R</strong> script that is loaded in the Source pane. First, lets see what directory we are in. To do so, type the following command into the script:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Since we will be learning several commands, we may already want to keep some short notes in our script to explain the purpose of the command. Entering a <code>#</code> before any line in an R script turns that line into a comment, which R will not try to interpret as code. Edit your script to include a comment on the purpose of commands you are learning, e.g.:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this command shows the current working directory</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<blockquote>
<h2 id="exercise-work-interactively-in-r">Exercise: Work interactively in R</h2>
<p>What happens when you try to enter the <code><a href="https://rdrr.io/r/base/getwd.html">getwd()</a></code> command in the Console pane?</p>
</blockquote>
<p>For the purposes of this exercise we want you to be in the directory <code>"/home/dcuser/R_data"</code>. What if you weren’t? You can set your home directory using the <code><a href="https://rdrr.io/r/base/getwd.html">setwd()</a></code> command. Enter this command in your script, but <em>don’t run</em> this yet.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This sets the working directory</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>You may have guessed, you need to tell the <code><a href="https://rdrr.io/r/base/getwd.html">setwd()</a></code> command what directory you want to set as your working directory. To do so, inside of the parentheses, open a set of quotes. Inside the quotes enter a <code><a href="https://rdrr.io/r/base/Arithmetic.html">/</a></code> which is the root directory for Linux. Next, use the <kbd>Tab</kbd> key, to take advantage of RStudio’s Tab-autocompletion method, to select <code>home</code>, <code>dcuser</code>, and <code>dc_genomics_r</code> directory. The path in your script should look like this:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This sets the working directory</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"/home/dcuser/dc_genomics_r"</span><span class="op">)</span></code></pre></div>
<p>When you run this command, the console repeats the command, but gives you no output. Instead, you see the blank R prompt: <code><a href="https://rdrr.io/r/base/Comparison.html">&gt;</a></code>. Congratulations! Although it seems small, knowing what your working directory is and being able to set your working directory is the first step to analyzing your data.</p>
<blockquote>
<h2 id="tip-never-use-setwd">Tip: Never use <code>setwd()</code>
</h2>
<p>Wait, what was the last 2 minutes about? Well, setting your working directory is something you need to do, you need to be very careful about using this as a step in your script. For example, what if your script is being on a computer that has a different directory structure? The top-level path in a Unix file system is root <code><a href="https://rdrr.io/r/base/Arithmetic.html">/</a></code>, but on Windows it is likely <code>C:\</code>. This is one of several ways you might cause a script to break because a file path is configured differently than your script anticipates. R packages like <a href="https://cran.r-project.org/package=here">here</a> and <a href="https://www.rdocumentation.org/packages/base/versions/3.4.3/topics/file.path">file.path</a> allow you to specify file paths is a way that is more operating system independent. See Jenny Bryan’s <a href="https://www.tidyverse.org/articles/2017/12/workflow-vs-script/">blog post</a> for this and other R tips.</p>
</blockquote>
</div>
<div id="using-functions-in-r-without-needing-to-master-them" class="section level2">
<h2 class="hasAnchor">
<a href="#using-functions-in-r-without-needing-to-master-them" class="anchor"></a>Using functions in R, without needing to master them</h2>
<p>A function in R (or any computing language) is a short program that takes some input and returns some output. Functions may seem like an advanced topic (and they are), but you have already used at least one function in R. <code><a href="https://rdrr.io/r/base/getwd.html">getwd()</a></code> is a function! The next sections will help you understand what is happening in any R script.</p>
<blockquote>
<h2 id="exercise-what-do-these-functions-do">Exercise: What do these functions do?</h2>
<p>Try the following functions by writing them in your script. See if you can guess what they do, and make sure to add comments to your script about your assumed purpose. - <code><a href="https://rdrr.io/r/base/list.files.html">dir()</a></code> - <code><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo()</a></code> - <code><a href="https://rdrr.io/r/base/date.html">date()</a></code> - <code><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time()</a></code></p>
</blockquote>
<p>You have hopefully noticed a pattern - an R function has three key properties: - Functions have a name (e.g. <code>dir</code>, <code>getwd</code>); note that functions are case sensitive! - Following the name, functions have a pair of <code>()</code> - Inside the parentheses, a function may take 0 or more arguments</p>
<p>An argument may be a specific input for your function and/or may modify the function’s behavior. For example the function <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> will round a number with a decimal:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This will round a number to the nearest integer</span>
<span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fl">3.14</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 3</code></pre>
</div>
<div id="getting-help-with-function-arguments" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-help-with-function-arguments" class="anchor"></a>Getting help with function arguments</h2>
<p>What if you wanted to round to one significant digit? <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> can do this, but you may first need to read the help to find out how. To see the help (In R sometimes also called a “vignette”) enter a <code><a href="https://rdrr.io/r/utils/Question.html">?</a></code> in front of the function name:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">?</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The “Help” tab will show you information (often, too much information). You will slowly learn how to read and make sense of help files. Checking the “Usage” or “Examples” headings is often a good place to look first. If you look under “Arguments,” we also see what arguments we can pass to this function to modify its behavior. You can also see a function’s argument using the <code><a href="https://rdrr.io/r/base/args.html">args()</a></code> function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/args.html">args</a></span><span class="op">(</span><span class="va">round</span><span class="op">)</span></code></pre></div>
<pre><code>## function (x, digits = 0) 
## NULL</code></pre>
<p><code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> takes two arguments, <code>x</code>, which is the number to be rounded, and a <code>digits</code> argument. The <code><a href="https://rdrr.io/r/base/assignOps.html">=</a></code> sign indicates that a default (in this case 0) is already set. Since <code>x</code> is not set, <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> requires we provide it, in contrast to <code>digits</code> where R will use the default value 0 unless you explicitly provide a different value. We can explicitly set the digits parameter when we call the function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fl">3.14159</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 3.14</code></pre>
<p>Or, R accepts what we call “positional arguments”, if you pass a function arguments separated by commas, R assumes that they are in the order you saw when we used <code><a href="https://rdrr.io/r/base/args.html">args()</a></code>. In the case below that means that <code>x</code> is 3.14159 and digits is 2.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fl">3.14159</span>, <span class="fl">2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 3.14</code></pre>
<p>Finally, what if you are using <code><a href="https://rdrr.io/r/utils/Question.html">?</a></code> to get help for a function in a package not installed on your system, such as when you are running a script which has dependencies.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">?</span><span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>will return an error:</p>
<pre><code>Error in .helpForCall(topicExpr, parent.frame()) :
   no methods for ‘geom_point’ and no documentation for it as a function</code></pre>
<p>Use two question marks (i.e. <code>??geom_point()</code>) and R will return results from a search of the documentation for packages you have installed on your computer in the “Help” tab. Finally, if you think there should be a function, for example a statistical test, but you aren’t sure what it is called in R, or what functions may be available, use the <code><a href="https://rdrr.io/r/utils/help.search.html">help.search()</a></code> function.</p>
<blockquote>
<h2 id="exercise-searching-for-r-functions">Exercise: Searching for R functions</h2>
<p>Use <code><a href="https://rdrr.io/r/utils/help.search.html">help.search()</a></code> to find R functions for the following statistical functions. Remember to put your search query in quotes inside the function’s parentheses.</p>
</blockquote>
<p>We will discuss more on where to look for the libraries and packages that contain functions you want to use. For now, be aware that two important ones are <a href="https://cran.r-project.org/">CRAN</a> - the main repository for R, and <a href="http://bioconductor.org/">Bioconductor</a> - a popular repository for bioinformatics-related R packages.</p>
</div>
<div id="rstudio-contextual-help" class="section level2">
<h2 class="hasAnchor">
<a href="#rstudio-contextual-help" class="anchor"></a>RStudio contextual help</h2>
<p>Here is one last bonus we will mention about RStudio. It’s difficult to remember all of the arguments and definitions associated with a given function. When you start typing the name of a function and hit the <kbd>Tab</kbd> key, RStudio will display functions and associated help:</p>
<p><img src="fig/studio_contexthelp1.png" alt="rstudio default session" style="width: 600px;"></p>
<p>Once you type a function, hitting the <kbd>Tab</kbd> inside the parentheses will show you the function’s arguments and provide additional help for each of these arguments.</p>
<p><img src="fig/studio_contexthelp2.png" alt="rstudio default session" style="width: 600px;"></p>
</div>
<div id="installing-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-packages" class="anchor"></a>Installing packages</h2>
<div id="from-cran-httpscran-r-project-org" class="section level3">
<h3 class="hasAnchor">
<a href="#from-cran-httpscran-r-project-org" class="anchor"></a>from CRAN (<a href="https://cran.r-project.org/" class="uri">https://cran.r-project.org/</a>)</h3>
<p>The Comprehensive R Archive network (CRAN) is a network of servers that stores R, documentation, and many of the packages available for R. To date there are over 170000 packages on CRAN, these packages can be installed by running the install.packages() command in an R terminal. For example, to install the <strong>ape</strong> package, run the following command in an R terminal:</p>
<pre><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("ape")</a></code></pre>
<p>To update all your installed packages just type:</p>
<pre><code><a href="https://rdrr.io/r/utils/update.packages.html">update.packages()</a></code></pre>
<p>Do not do it now!!</p>
</div>
<div id="from-bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#from-bioconductor" class="anchor"></a>from Bioconductor</h3>
<p>Bioconductor is another archive of R packages specific to bioinformatics and genomics. This archive is maintained by the Bioconductor core team and is updated bi-annually.</p>
<pre><code># Install biocManager if it's not already there
if (!requireNamespace("BiocManager"))
    install.packages("BiocManager") # from CRAN

# Install core bioconductor packages
BiocManager::install()

# Install specific bioconductor packages
BiocManager::install("GenomicFeatures")

# Upgrade installed bioconductor packages
BiocManager::install()</code></pre>
</div>
<div id="from-github" class="section level3">
<h3 class="hasAnchor">
<a href="#from-github" class="anchor"></a>from github</h3>
<p>Finally there are lots of packages on github. These are often new pacages or development versions.</p>
<pre><code># Install the remotes package if it's not already there
if (!requireNamespace("remotes"))
    install.packages("remotes") 

remotes::install_github("KlausVigo/phangorn")</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Klaus Schliep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
