<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>R Basics • Metagenomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="R Basics">
<meta property="og:description" content="Metagenomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Metagenomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01-introduction.html">Introducing R and RStudio</a>
    </li>
    <li>
      <a href="../articles/02-r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="../articles/03-basics-factors-dataframes.html">R Basics continued - factors and data frames</a>
    </li>
    <li>
      <a href="../articles/04-dplyr.html">Aggregating and Analyzing Data with dplyr</a>
    </li>
    <li>
      <a href="../articles/05-data-visualization.html">Data Visualization with ggplot2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="02-r-basics_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>R Basics</h1>
                        <h4 class="author">Klaus Schliep</h4>
            
            <h4 class="date">Mai 17, 2021</h4>
      
      
      <div class="hidden name"><code>02-r-basics.Rmd</code></div>

    </div>

    
    
<div id="the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r" class="section level2">
<h2 class="hasAnchor">
<a href="#the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r" class="anchor"></a>“The fantastic world of R awaits you” OR “Nobody wants to learn how to use R”</h2>
<p>Before we begin this lesson, we want you to be clear on the goal of the workshop and these lessons. We believe that every learner can <strong>achieve competency with R</strong>. You have reached competency when you find that you are able to <strong>use R to handle common analysis challenges in a reasonable amount of time</strong> (which includes time needed to look at learning materials, search for answers online, and ask colleagues for help). As you spend more time using R (there is no substitute for regular use and practice) you will find yourself gaining competency and even expertise. The more familiar you get, the more complex the analyses you will be able to carry out, with less frustration, and in less time - the fantastic world of R awaits you!</p>
</div>
<div id="what-these-lessons-will-not-teach-you" class="section level2">
<h2 class="hasAnchor">
<a href="#what-these-lessons-will-not-teach-you" class="anchor"></a>What these lessons will not teach you</h2>
<p>Nobody wants to learn how to use R. People want to learn how to use R to analyze their own research questions! Ok, maybe some folks learn R for R’s sake, but these lessons assume that you want to start analyzing genomic data as soon as possible. Given this, there are many valuable pieces of information about R that we simply won’t have time to cover. Hopefully, we will clear the hurdle of giving you just enough knowledge to be dangerous, which can be a high bar in R! We suggest you look into the additional learning materials in the tip box below.</p>
<p><strong>Here are some R skills we will <em>not</em> cover in these lessons</strong></p>
<ul>
<li>How to create and work with R matrices and R lists</li>
<li>How to create and work with loops and conditional statements, and the “apply” family of functions (which are super useful, read more <a href="https://www.r-bloggers.com/r-tutorial-on-the-apply-family-of-functions/">here</a>)</li>
<li>How to do basic string manipulations (e.g. finding patterns in text using grep, replacing text)</li>
<li>How to plot using the default R graphic tools (we <em>will</em> cover plot creation, but will do so using the popular plotting package <code>ggplot2</code>)</li>
<li>How to use advanced R statistical functions</li>
</ul>
<blockquote>
<h2 id="tip-where-to-learn-more">Tip: Where to learn more</h2>
<p>The following are good resources for learning more about R. Some of them can be quite technical, but if you are a regular R user you may ultimately need this technical knowledge. - <a href="https://cran.r-project.org/doc/contrib/Paradis-rdebuts_en.pdf">R for Beginners</a>: By Emmanuel Paradis and a great starting point - <a href="https://cran.r-project.org/manuals.html">The R Manuals</a>: Maintained by the R project - <a href="https://cran.r-project.org/other-docs.html">R contributed documentation</a>: Also linked to the R project; importantly there are materials available in several languages - <a href="http://r4ds.had.co.nz/">R for Data Science</a>: A wonderful collection by noted R educators and developers Garrett Grolemund and Hadley Wickham - <a href="https://peerj.com/collections/50-practicaldatascistats/">Practical Data Science for Stats</a>: Not exclusively about R usage, but a nice collection of pre-prints on data science and applications for R - <a href="https://software-carpentry.org/lessons/">Programming in R Software Carpentry lesson</a>: There are several Software Carpentry lessons in R to choose from {: .callout}</p>
</blockquote>
</div>
<div id="creating-objects-in-r" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-objects-in-r" class="anchor"></a>Creating objects in R</h2>
<blockquote>
<h2 id="reminder">Reminder</h2>
<p>At this point you should be coding along in the “<strong>genomics_r_basics.R</strong>” script we created in the last episode. Writing your commands in the script (and commenting it) will make it easier to record what you did and why. {: .prereq}</p>
</blockquote>
<p>What might be called a variable in many languages is called an <strong>object</strong> in R.</p>
<p><strong>To create an object you need:</strong></p>
<ul>
<li>a name (e.g. ‘a’)</li>
<li>a value (e.g. ‘1’)</li>
<li>the assignment operator (‘&lt;-’)</li>
</ul>
<p>In your script, “<strong>genomics_r_basics.R</strong>”, using the R assignment operator ‘&lt;-’, assign ‘1’ to the object ‘a’ as shown. Remember to leave a comment in the line above (using the ‘#’) to explain what you are doing:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this line creates the object 'a' and assigns it the value '1'</span>

<span class="va">a</span> <span class="op">&lt;-</span> <span class="fl">1</span></code></pre></div>
<p>Next, run this line of code in your script. You can run a line of code by hitting the <kbd>Run</kbd> button that is just above the first line of your script in the header of the Source pane or you can use the appropriate shortcut:</p>
<ul>
<li>Windows execution shortcut: <kbd>Ctrl</kbd>+<kbd>Enter</kbd>
</li>
<li>Mac execution shortcut: <kbd>Cmd(⌘)</kbd>+<kbd>Enter</kbd>
</li>
</ul>
<p>To run multiple lines of code, you can highlight all the line you wish to run and then hit <kbd>Run</kbd> or use the shortcut key combo listed above.</p>
<p>In the RStudio ‘Console’ you should see:</p>
<pre><code>a &lt;- 1
&gt;</code></pre>
<p>{: .output}</p>
<p>The ‘Console’ will display lines of code run from a script and any outputs or status/warning/error messages (usually in red).</p>
<p>In the ‘Environment’ window you will also get a table:</p>
<table class="table">
<thead><tr class="header">
<th>Values</th>
<th></th>
</tr></thead>
<tbody><tr class="odd">
<td>a</td>
<td>1</td>
</tr></tbody>
</table>
<p>The ‘Environment’ window allows you to keep track of the objects you have created in R.</p>
<blockquote>
<h2 id="exercise-create-some-objects-in-r">Exercise: Create some objects in R</h2>
<p>Create the following objects; give each object an appropriate name (your best guess at what name to use is fine):</p>
<ol style="list-style-type: decimal">
<li>Create an object that has the value of number of pairs of human chromosomes</li>
<li>Create an object that has a value of your favorite gene name</li>
<li>Create an object that has this URL as its value: “<a href="ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/" class="uri">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/</a>”</li>
<li>Create an object that has the value of the number of chromosomes in a diploid human cell</li>
</ol>
</blockquote>
</div>
<div id="naming-objects-in-r" class="section level2">
<h2 class="hasAnchor">
<a href="#naming-objects-in-r" class="anchor"></a>Naming objects in R</h2>
<p>Here are some important details about naming objects in R.</p>
<ul>
<li>
<strong>Avoid spaces and special characters</strong>: Object names cannot contain spaces or the minus sign (<code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code>). You can use ’_’ to make names more readable. You should avoid using special characters in your object name (e.g. ! @ # . , etc.). Also, object names cannot begin with a number.</li>
<li>
<strong>Use short, easy-to-understand names</strong>: You should avoid naming your objects using single letters (e.g. ‘n’, ‘p’, etc.). This is mostly to encourage you to use names that would make sense to anyone reading your code (a colleague, or even yourself a year from now). Also, avoiding excessively long names will make your code more readable.</li>
<li>
<strong>Avoid commonly used names</strong>: There are several names that may already have a definition in the R language (e.g. ‘mean’, ‘min’, ‘max’). One clue that a name already has meaning is that if you start typing a name in RStudio and it gets a colored highlight or RStudio gives you a suggested autocompletion you have chosen a name that has a reserved meaning.</li>
<li>
<strong>Use the recommended assignment operator</strong>: In R, we use ‘&lt;-’ as the preferred assignment operator. ‘=’ works too, but is most commonly used in passing arguments to functions (more on functions later). There is a shortcut for the R assignment operator:
<ul>
<li>Windows execution shortcut: <kbd>Alt</kbd>+<kbd>-</kbd>
</li>
<li>Mac execution shortcut: <kbd>Option</kbd>+<kbd>-</kbd>
</li>
</ul>
</li>
</ul>
<p>There are a few more suggestions about naming and style you may want to learn more about as you write more R code. There are several “style guides” that have advice, and one to start with is the <a href="http://style.tidyverse.org/index.html">tidyverse R style guide</a>.</p>
<blockquote>
<h2 id="tip-pay-attention-to-warnings-in-the-script-console">Tip: Pay attention to warnings in the script console</h2>
<p>If you enter a line of code in your script that contains an error, RStudio may give you an error message and underline this mistake. Sometimes these messages are easy to understand, but often the messages may need some figuring out. Paying attention to these warnings will help you avoid mistakes. In the example below, our object name has a space, which is not allowed in R. The error message does not say this directly, but R is “not sure” about how to assign the name to “human_ chr_number” when the object name we want is “human_chr_number”.</p>
<p><img src="fig/rstudio_script_warning.png" alt="rstudio script warning" style="width: 600px;"></p>
</blockquote>
</div>
<div id="reassigning-object-names-or-deleting-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#reassigning-object-names-or-deleting-objects" class="anchor"></a>Reassigning object names or deleting objects</h2>
<p>Once an object has a value, you can change that value by overwriting it. R will not give you a warning or error if you overwriting an object, which may or may not be a good thing depending on how you look at it.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># gene_name has the value 'pten' or whatever value you used in the challenge.</span>
<span class="co"># We will now assign the new value 'tp53'</span>
<span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="st">'tp53'</span></code></pre></div>
<p>You can also remove an object from R’s memory entirely. The <code><a href="https://rdrr.io/r/base/rm.html">rm()</a></code> function will delete the object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># delete the object 'gene_name'</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span></code></pre></div>
<p>If you run a line of code that has only an object name, R will normally display the contents of that object. In this case, we are told the object no longer exists.</p>
<pre><code>Error: object 'gene_name' not found</code></pre>
<p>{: .error}</p>
</div>
<div id="understanding-object-data-types-modes" class="section level2">
<h2 class="hasAnchor">
<a href="#understanding-object-data-types-modes" class="anchor"></a>Understanding object data types (modes)</h2>
<p>In R, <strong>every object has two properties</strong>:</p>
<ul>
<li>
<strong>Length</strong>: How many distinct values are held in that object</li>
<li>
<strong>Mode</strong>: What is the classification (type) of that object.</li>
</ul>
<p>We will get to the “length” property later in the lesson. The <strong>“mode” property</strong> <strong>corresponds to the type of data an object represents</strong>. The most common modes you will encounter in R are:</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="75%">
</colgroup>
<thead><tr class="header">
<th>Mode (abbreviation)</th>
<th>Type of data</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Numeric (num)</td>
<td>Numbers such floating point/decimals (1.0, 0.5, 3.14), there are also more specific numeric types (dbl - Double, int - Integer). These differences are not relevant for most beginners and pertain to how these values are stored in memory</td>
</tr>
<tr class="even">
<td>Character (chr)</td>
<td>A sequence of letters/numbers in single ’’ or double " " quotes</td>
</tr>
<tr class="odd">
<td>Logical</td>
<td>Boolean values - TRUE or FALSE</td>
</tr>
</tbody>
</table>
<p>There are a few other modes (i.e. “complex”, “raw” etc.) but these are the three we will work with in this lesson.</p>
<p>Data types are familiar in many programming languages, but also in natural language where we refer to them as the parts of speech, e.g. nouns, verbs, adverbs, etc. Once you know if a word - perhaps an unfamiliar one - is a noun, you can probably guess you can count it and make it plural if there is more than one (e.g. 1 <a href="https://en.wikipedia.org/wiki/Tuatara">Tuatara</a>, or 2 Tuataras). If something is a adjective, you can usually change it into an adverb by adding “-ly” (e.g. <a href="https://www.merriam-webster.com/dictionary/jejune">jejune</a> vs. jejunely). Depending on the context, you may need to decide if a word is in one category or another (e.g “cut” may be a noun when it’s on your finger, or a verb when you are preparing vegetables). These concepts have important analogies when working with R objects.</p>
<blockquote>
<h2 id="exercise-create-objects-and-check-their-modes">Exercise: Create objects and check their modes</h2>
<p>Create the following objects in R, then use the <code><a href="https://rdrr.io/r/base/mode.html">mode()</a></code> function to verify their modes. Try to guess what the mode will be before you look at the solution</p>
<ol style="list-style-type: decimal">
<li><code>chromosome_name &lt;- 'chr02'</code></li>
<li><code>od_600_value &lt;- 0.47</code></li>
<li><code>chr_position &lt;- '1001701'</code></li>
<li><code>spock &lt;- TRUE</code></li>
<li><code>pilot &lt;- Earhart</code></li>
</ol>
</blockquote>
<p>Notice from the solution that even if a series of numbers is given as a value R will consider them to be in the “character” mode if they are enclosed as single or double quotes. Also, notice that you cannot take a string of alphanumeric characters (e.g. Earhart) and assign as a value for an object. In this case, R looks for an object named <code>Earhart</code> but since there is no object, no assignment can be made. If <code>Earhart</code> did exist, then the mode of <code>pilot</code> would be whatever the mode of <code>Earhart</code> was originally. If we want to create an object called <code>pilot</code> that was the <strong>name</strong> “Earhart”, we need to enclose <code>Earhart</code> in quotation marks.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pilot</span> <span class="op">&lt;-</span> <span class="st">"Earhart"</span>
<span class="fu"><a href="https://rdrr.io/r/base/mode.html">mode</a></span><span class="op">(</span><span class="va">pilot</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
</div>
<div id="mathematical-and-functional-operations-on-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#mathematical-and-functional-operations-on-objects" class="anchor"></a>Mathematical and functional operations on objects</h2>
<p>Once an object exists (which by definition also means it has a mode), R can appropriately manipulate that object. For example, objects of the numeric modes can be added, multiplied, divided, etc. R provides several mathematical (arithmetic) operators including:</p>
<table class="table">
<thead><tr class="header">
<th>Operator</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>+</td>
<td>addition</td>
</tr>
<tr class="even">
<td>-</td>
<td>subtraction</td>
</tr>
<tr class="odd">
<td>*</td>
<td>multiplication</td>
</tr>
<tr class="even">
<td>/</td>
<td>division</td>
</tr>
<tr class="odd">
<td>^ or **</td>
<td>exponentiation</td>
</tr>
<tr class="even">
<td>a%/%b</td>
<td>integer division (division where the remainder is discarded)</td>
</tr>
<tr class="odd">
<td>a%%b</td>
<td>modulus (returns the remainder after division)</td>
</tr>
</tbody>
</table>
<p>These can be used with literal numbers:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="op">(</span><span class="fl">5</span> <span class="op">**</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></code></pre></div>
<pre><code>## [1] 1.618034</code></pre>
<p>and importantly, can be used on any object that evaluates to (i.e. interpreted by R) a numeric object:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># multiply the object 'human_chr_number' by 2</span>

<span class="va">human_chr_number</span> <span class="op">*</span> <span class="fl">2</span></code></pre></div>
<pre><code>## [1] 46</code></pre>
<blockquote>
<h2 id="exercise-compute-the-golden-ratio">Exercise: Compute the golden ratio</h2>
<p>One approximation of the golden ratio (φ) can be found by taking the sum of 1 and the square root of 5, and dividing by 2 as in the example above. Compute the golden ratio to 3 digits of precision using the <code><a href="https://rdrr.io/r/base/MathFun.html">sqrt()</a></code> and <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> functions. Hint: remember the <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code> function can take 2 arguments.</p>
</blockquote>
</div>
<div id="vectors" class="section level2">
<h2 class="hasAnchor">
<a href="#vectors" class="anchor"></a>Vectors</h2>
<p>Vectors are probably the most used commonly used object type in R. <strong>A vector is a collection of values that are all of the same type (numbers, characters, etc.)</strong>. One of the most common ways to create a vector is to use the <code><a href="https://rdrr.io/r/base/c.html">c()</a></code> function - the “concatenate” or “combine” function. Inside the function you may enter one or more values; for multiple values, separate each value with a comma:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create the SNP gene name vector</span>

<span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"OXTR"</span>, <span class="st">"ACTN3"</span>, <span class="st">"AR"</span>, <span class="st">"OPRM1"</span><span class="op">)</span></code></pre></div>
<p>Vectors always have a <strong>mode</strong> and a <strong>length</strong>. You can check these with the <code><a href="https://rdrr.io/r/base/mode.html">mode()</a></code> and <code><a href="https://rdrr.io/r/base/length.html">length()</a></code> functions respectively. Another useful function that gives both of these pieces of information is the <code><a href="https://rdrr.io/r/utils/str.html">str()</a></code> (structure) function.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Check the mode, length, and structure of 'snp_genes'</span>
<span class="fu"><a href="https://rdrr.io/r/base/mode.html">mode</a></span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></code></pre></div>
<pre><code>##  chr [1:4] "OXTR" "ACTN3" "AR" "OPRM1"</code></pre>
<p>Vectors are quite important in R. Another data type that we will work with later in this lesson, data frames, are collections of vectors. What we learn here about vectors will pay off even more when we start working with data frames.</p>
</div>
<div id="creating-and-subsetting-vectors" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-and-subsetting-vectors" class="anchor"></a>Creating and subsetting vectors</h2>
<p>Let’s create a few more vectors to play around with:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Some interesting human SNPs</span>
<span class="co"># while accuracy is important, typos in the data won't hurt you here</span>

<span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'rs53576'</span>, <span class="st">'rs1815739'</span>, <span class="st">'rs6152'</span>, <span class="st">'rs1799971'</span><span class="op">)</span>
<span class="va">snp_chromosomes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'3'</span>, <span class="st">'11'</span>, <span class="st">'X'</span>, <span class="st">'6'</span><span class="op">)</span>
<span class="va">snp_positions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8762685</span>, <span class="fl">66560624</span>, <span class="fl">67545785</span>, <span class="fl">154039662</span><span class="op">)</span></code></pre></div>
<p>Once we have vectors, one thing we may want to do is specifically retrieve one or more values from our vector. To do so, we use <strong>bracket notation</strong>. We type the name of the vector followed by square brackets. In those square brackets we place the index (e.g. a number) in that bracket as follows:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the 3rd value in the snp_genes vector</span>
<span class="va">snp_genes</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "AR"</code></pre>
<p>In R, every item your vector is indexed, starting from the first item (1) through to the final number of items in your vector. You can also retrieve a range of numbers:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the 1st through 3rd value in the snp_genes vector</span>

<span class="va">snp_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "OXTR"  "ACTN3" "AR"</code></pre>
<p>If you want to retrieve several (but not necessarily sequential) items from a vector, you pass a <strong>vector of indices</strong>; a vector that has the numbered positions you wish to retrieve.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the 1st, 3rd, and 4th value in the snp_genes vector</span>

<span class="va">snp_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "OXTR"  "AR"    "OPRM1"</code></pre>
<p>There are additional (and perhaps less commonly used) ways of subsetting a vector (see <a href="https://thomasleeper.com/Rcourse/Tutorials/vectorindexing.html">these examples</a>). Also, several of these subsetting expressions can be combined:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the 1st through the 3rd value, and 4th value in the snp_genes vector</span>
<span class="co"># yes, this is a little silly in a vector of only 4 values.</span>
<span class="va">snp_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "OXTR"  "ACTN3" "AR"    "OPRM1"</code></pre>
</div>
<div id="adding-to-removing-or-replacing-values-in-existing-vectors" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-to-removing-or-replacing-values-in-existing-vectors" class="anchor"></a>Adding to, removing, or replacing values in existing vectors</h2>
<p>Once you have an existing vector, you may want to add a new item to it. To do so, you can use the <code><a href="https://rdrr.io/r/base/c.html">c()</a></code> function again to add your new value:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># add the gene 'CYP1A1' and 'APOA5' to our list of snp genes</span>
<span class="co"># this overwrites our existing vector</span>
<span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">snp_genes</span>, <span class="st">"CYP1A1"</span>, <span class="st">"APOA5"</span><span class="op">)</span></code></pre></div>
<p>We can verify that “snp_genes” contains the new gene entry</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_genes</span></code></pre></div>
<pre><code>## [1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1" "APOA5"</code></pre>
<p>Using a negative index will return a version of a vector with that index’s value removed:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_genes</span><span class="op">[</span><span class="op">-</span><span class="fl">6</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1"</code></pre>
<p>We can remove that value from our vector by overwriting it with this expression:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="va">snp_genes</span><span class="op">[</span><span class="op">-</span><span class="fl">6</span><span class="op">]</span>
<span class="va">snp_genes</span></code></pre></div>
<pre><code>## [1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1"</code></pre>
<p>We can also explicitly rename or add a value to our index using double bracket notation:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_genes</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">&lt;-</span> <span class="st">"APOA5"</span>
<span class="va">snp_genes</span></code></pre></div>
<pre><code>## [1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1" NA       "APOA5"</code></pre>
<p>Notice in the operation above that R inserts an <code>NA</code> value to extend our vector so that the gene “APOA5” is an index 7. This may be a good or not-so-good thing depending on how you use this.</p>
<blockquote>
<h2 id="exercise-examining-and-subsetting-vectors">Exercise: Examining and subsetting vectors</h2>
<p>Answer the following questions to test your knowledge of vectors</p>
<p>Which of the following are true of vectors in R? A) All vectors have a mode <strong>or</strong> a length<br>
B) All vectors have a mode <strong>and</strong> a length<br>
C) Vectors may have different lengths<br>
D) Items within a vector may be of different modes<br>
E) You can use the <code><a href="https://rdrr.io/r/base/c.html">c()</a></code> to one or more items to an existing vector<br>
F) You can use the <code><a href="https://rdrr.io/r/base/c.html">c()</a></code> to add a vector to an exiting vector<br>
&gt;</p>
</blockquote>
</div>
<div id="logical-subsetting" class="section level2">
<h2 class="hasAnchor">
<a href="#logical-subsetting" class="anchor"></a>Logical Subsetting</h2>
<p>There is one last set of cool subsetting capabilities we want to introduce. It is possible within R to retrieve items in a vector based on a logical evaluation or numerical comparison. For example, let’s say we wanted get all of the SNPs in our vector of SNP positions that were greater than 100,000,000. We could index using the ‘&gt;’ (greater than) logical operator:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions</span><span class="op">[</span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="fl">100000000</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 154039662</code></pre>
<p>In the square brackets you place the name of the vector followed by the comparison operator and (in this case) a numeric value. Some of the most common logical operators you will use in R are:</p>
<table class="table">
<thead><tr class="header">
<th>Operator</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>&lt;</td>
<td>less than</td>
</tr>
<tr class="even">
<td>&lt;=</td>
<td>less than or equal to</td>
</tr>
<tr class="odd">
<td>&gt;</td>
<td>greater than</td>
</tr>
<tr class="even">
<td>&gt;=</td>
<td>greater than or equal to</td>
</tr>
<tr class="odd">
<td>==</td>
<td>exactly equal to</td>
</tr>
<tr class="even">
<td>!=</td>
<td>not equal to</td>
</tr>
<tr class="odd">
<td>!x</td>
<td>not x</td>
</tr>
<tr class="even">
<td>a | b</td>
<td>a or b</td>
</tr>
<tr class="odd">
<td>a &amp; b</td>
<td>a and b</td>
</tr>
</tbody>
</table>
<blockquote>
<h2 id="the-magic-of-programming">The magic of programming</h2>
<p>The reason why the expression <code>snp_positions[snp_positions &gt; 100000000]</code> works can be better understood if you examine what the expression “snp_positions &gt; 100000000” evaluates to:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="fl">100000000</span></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE  TRUE</code></pre>
<p>The output above is a logical vector, the 4th element of which is TRUE. When you pass a logical vector as an index, R will return the true values:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 154039662</code></pre>
<p>If you have never coded before, this type of situation starts to expose the “magic” of programming. We mentioned before that in the bracket notation you take your named vector followed by brackets which contain an index: <strong>named_vector[index]</strong>. The “magic” is that the index needs to <em>evaluate to</em> a number. So, even if it does not appear to be an integer (e.g. 1, 2, 3), as long as R can evaluate it, we will get a result. That our expression <code>snp_positions[snp_positions &gt; 100000000]</code> evaluates to a number can be seen in the following situation. If you wanted to know which <strong>index</strong> (1, 2, 3, or 4) in our vector of SNP positions was the one that was greater than 100,000,000?</p>
<p>We can use the <code><a href="https://rdrr.io/r/base/which.html">which()</a></code> function to return the indices of any item that evaluates as TRUE in our comparison:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="fl">100000000</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<p><strong>Why this is important</strong></p>
<p>Often in programming we will not know what inputs and values will be used when our code is executed. Rather than put in a pre-determined value (e.g 100000000) we can use an object that can take on whatever value we need. So for example:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_marker_cutoff</span> <span class="op">&lt;-</span> <span class="fl">100000000</span>
<span class="va">snp_positions</span><span class="op">[</span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="va">snp_marker_cutoff</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 154039662</code></pre>
<p>Ultimately, it’s putting together flexible, reusable code like this that gets at the “magic” of programming! {: .callout}</p>
</blockquote>
</div>
<div id="a-few-final-vector-tricks" class="section level2">
<h2 class="hasAnchor">
<a href="#a-few-final-vector-tricks" class="anchor"></a>A few final vector tricks</h2>
<p>Finally, there are a few other common retrieve or replace operations you may want to know about. First, you can check to see if any of the values of your vector are missing (i.e. are <code>NA</code>). Missing data will get a more detailed treatment later, but the <code>is.NA()</code> function will return a logical vector, with TRUE for any NA value:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># current value of 'snp_genes': </span>
<span class="co"># chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</span>

<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE</code></pre>
<p>Sometimes, you may wish to find out if a specific value (or several values) is present a vector. You can do this using the comparison operator <code><a href="https://rdrr.io/r/base/match.html">%in%</a></code>, which will return TRUE for any value in your collection that is in the vector you are searching:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># current value of 'snp_genes':</span>
<span class="co"># chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</span>

<span class="co"># test to see if "ACTN3" or "APO5A" is in the snp_genes vector</span>
<span class="co"># if you are looking for more than one value, you must pass this as a vector</span>

<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ACTN3"</span>,<span class="st">"APOA5"</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">snp_genes</span></code></pre></div>
<pre><code>## [1] TRUE TRUE</code></pre>
<blockquote>
<h2 id="review-exercise-1">Review Exercise 1</h2>
<p>What data types/modes are the following vectors? a. <code>snps</code><br>
b. <code>snp_chromosomes</code><br>
c. <code>snp_positions</code></p>
</blockquote>
<blockquote>
<h2 id="review-exercise-2">Review Exercise 2</h2>
<p>Add the following values to the specified vectors: a. To the <code>snps</code> vector add: ‘rs662799’<br>
b. To the <code>snp_chromosomes</code> vector add: 11<br>
c. To the <code>snp_positions</code> vector add: 116792991</p>
</blockquote>
<blockquote>
<h2 id="review-exercise-3">Review Exercise 3</h2>
<p>Make the following change to the <code>snp_genes</code> vector:</p>
<p>Hint: Your vector should look like this in ‘Environment’: <code>chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</code>. If not recreate the vector by running this expression: <code>snp_genes &lt;- c("OXTR", "ACTN3", "AR", "OPRM1", "CYP1A1", NA, "APOA5")</code></p>
<ol style="list-style-type: lower-alpha">
<li>Create a new version of <code>snp_genes</code> that does not contain CYP1A1 and then<br>
</li>
<li>Add 2 NA values to the end of <code>snp_genes</code>
</li>
</ol>
</blockquote>
<blockquote>
<h2 id="review-exercise-4">Review Exercise 4</h2>
<p>Using indexing, create a new vector named <code>combined</code> that contains: - The the 1st value in <code>snp_genes</code> - The 1st value in <code>snps</code> - The 1st value in <code>snp_chromosomes</code> - The 1st value in <code>snp_positions</code></p>
</blockquote>
<blockquote>
<h2 id="review-exercise-5">Review Exercise 5</h2>
<p>What type of data is <code>combined</code>?</p>
</blockquote>
</div>
<div id="bonus-material-lists" class="section level2">
<h2 class="hasAnchor">
<a href="#bonus-material-lists" class="anchor"></a>Bonus material: Lists</h2>
<p>Lists are quite useful in R, but we won’t be using them in the genomics lessons. That said, you may come across lists in the way that some bioinformatics programs may store and/or return data to you. One of the key attributes of a list is that, unlike a vector, a list may contain data of more than one mode. Learn more about creating and using lists using this <a href="https://r4ds.had.co.nz/vectors.html#lists">nice tutorial</a>. In this one example, we will create a named list and show you how to retrieve items from the list.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create a named list using the 'list' function and our SNP examples</span>
<span class="co"># Note, for easy reading we have placed each item in the list on a separate line</span>
<span class="co"># Nothing special about this, you can do this for any multiline commands</span>
<span class="co"># To run this command, make sure the entire command (all 4 lines) are highlighted</span>
<span class="co"># before running</span>
<span class="co"># Note also, as we are doing all this inside the list() function use of the</span>
<span class="co"># '=' sign is good style</span>
<span class="va">snp_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="va">snp_genes</span>,
                 refference_snp <span class="op">=</span> <span class="va">snps</span>,
                 chromosome <span class="op">=</span> <span class="va">snp_chromosomes</span>,
                 position <span class="op">=</span> <span class="va">snp_positions</span><span class="op">)</span>
<span class="co"># Examine the structure of the list</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">snp_data</span><span class="op">)</span></code></pre></div>
<pre><code>## List of 4
##  $ genes         : chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" ...
##  $ refference_snp: chr [1:4] "rs53576" "rs1815739" "rs6152" "rs1799971"
##  $ chromosome    : chr [1:4] "3" "11" "X" "6"
##  $ position      : num [1:4] 8.76e+06 6.66e+07 6.75e+07 1.54e+08</code></pre>
<p>To get all the values for the <code>position</code> object in the list, we use the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> notation:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># return all the values of position object</span>

<span class="va">snp_data</span><span class="op">$</span><span class="va">position</span></code></pre></div>
<pre><code>## [1]   8762685  66560624  67545785 154039662</code></pre>
<p>To get the first value in the <code>position</code> object, use the <code>[]</code> notation to index:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># return first value of the position object</span>

<span class="va">snp_data</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 8762685</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Klaus Schliep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
