<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>R Basics continued - factors and data frames • Metagenomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="R Basics continued - factors and data frames">
<meta property="og:description" content="Metagenomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Metagenomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01-introduction.html">Introducing R and RStudio</a>
    </li>
    <li>
      <a href="../articles/02-r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="../articles/03-basics-factors-dataframes.html">R Basics continued - factors and data frames</a>
    </li>
    <li>
      <a href="../articles/04-dplyr.html">Aggregating and Analyzing Data with dplyr</a>
    </li>
    <li>
      <a href="../articles/05-data-visualization.html">Data Visualization with ggplot2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="03-basics-factors-dataframes_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>R Basics continued - factors and data frames</h1>
                        <h4 class="author">Klaus Schliep</h4>
            
            <h4 class="date">Mai 17, 2021</h4>
      
      
      <div class="hidden name"><code>03-basics-factors-dataframes.Rmd</code></div>

    </div>

    
    
<div id="working-with-spreadsheets-tabular-data" class="section level2">
<h2 class="hasAnchor">
<a href="#working-with-spreadsheets-tabular-data" class="anchor"></a>Working with spreadsheets (tabular data)</h2>
<p>A substantial amount of the data we work with in genomics will be tabular data, this is data arranged in rows and columns - also known as spreadsheets. We could write a whole lesson on how to work with spreadsheets effectively (<a href="https://datacarpentry.org/organization-genomics/">actually we did</a>). For our purposes, we want to remind you of a few principles before we work with our first set of example data:</p>
<p><strong>1) Keep raw data separate from analyzed data</strong></p>
<p>This is principle number one because if you can’t tell which files are the original raw data, you risk making some serious mistakes (e.g. drawing conclusion from data which have been manipulated in some unknown way).</p>
<p><strong>2) Keep spreadsheet data Tidy</strong></p>
<p>The simplest principle of <strong>Tidy data</strong> is that we have one row in our spreadsheet for each observation or sample, and one column for every variable that we measure or report on. As simple as this sounds, it’s very easily violated. Most data scientists agree that significant amounts of their time is spent tidying data for analysis. Read more about data organization in <a href="https://datacarpentry.org/organization-genomics/">our lesson</a> and in <a href="https://www.jstatsoft.org/article/view/v059i10">this paper</a>.</p>
<p><strong>3) Trust but verify</strong></p>
<p>Finally, while you don’t need to be paranoid about data, you should have a plan for how you will prepare it for analysis. <strong>This a focus of this lesson.</strong> You probably already have a lot of intuition, expectations, assumptions about your data - the range of values you expect, how many values should have been recorded, etc. Of course, as the data get larger our human ability to keep track will start to fail (and yes, it can fail for small data sets too). R will help you to examine your data so that you can have greater confidence in your analysis, and its reproducibility.</p>
<blockquote>
<h2 id="tip-keeping-you-raw-data-separate">Tip: Keeping you raw data separate</h2>
<p>When you work with data in R, you are not changing the original file you loaded that data from. This is different than (for example) working with a spreadsheet program where changing the value of the cell leaves you one “save”-click away from overwriting the original file. You have to purposely use a writing function (e.g. <code><a href="https://rdrr.io/r/utils/write.table.html">write.csv()</a></code>) to save data loaded into R. In that case, be sure to save the manipulated data into a new file. More on this later in the lesson.</p>
</blockquote>
</div>
<div id="importing-tabular-data-into-r" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-tabular-data-into-r" class="anchor"></a>Importing tabular data into R</h2>
<p>There are several ways to import data into R. For our purpose here, we will focus on using the tools every R installation comes with (so called “base” R) to import a comma-delimited file containing the results of our variant calling workflow. We will need to load the sheet using a function called <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code>.</p>
<blockquote>
<h2 id="exercise-review-the-arguments-of-the-read.csv-function">Exercise: Review the arguments of the <code>read.csv()</code> function</h2>
<p><strong>Before using the <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> function, use R’s help feature to answer the following questions</strong>.</p>
<p><em>Hint</em>: Entering ‘?’ before the function name and then running that line will bring up the help documentation. Also, when reading this particular help be careful to pay attention to the ‘read.csv’ expression under the ‘Usage’ heading. Other answers will be in the ‘Arguments’ heading.</p>
<ol style="list-style-type: upper-alpha">
<li><p>What is the default parameter for ‘header’ in the <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> function?</p></li>
<li><p>What argument would you have to change to read a file that was delimited by semicolons (;) rather than commas?</p></li>
<li><p>What argument would you have to change to read file in which numbers used commas for decimal separation (i.e. 1,00)?</p></li>
<li><p>What argument would you have to change to read in only the first 10,000 rows of a very large file?</p></li>
</ol>
</blockquote>
<p>Now, let’s read in the file <code>combined_tidy_vcf.csv</code> which will be located in <code>https://github.com/KlausVigo/Metagenomics/tree/master/vignettes/combined_tidy_vcf.csv</code>. Call this data <code>variants</code>. The first argument to pass to our <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> function is the file path for our data. The file path must be in quotes and now is a good time to remember to use tab autocompletion. <strong>If you use tab autocompletion you avoid typos and errors in file paths.</strong> Use it!</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## read in a CSV file and save it as 'variants'</span>

<span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"combined_tidy_vcf.csv"</span><span class="op">)</span></code></pre></div>
<p>One of the first things you should notice is that in the Environment window, you have the <code>variants</code> object, listed as 801 obs. (observations/rows) of 29 variables (columns). Double-clicking on the name of the object will open a view of the data in a new tab.</p>
<p><img src="fig/rstudio_dataframeview.png" alt="rstudio data frame view" style="width: 1000px;"></p>
</div>
<div id="summarizing-and-determining-the-structure-of-a-data-frame-" class="section level2">
<h2 class="hasAnchor">
<a href="#summarizing-and-determining-the-structure-of-a-data-frame-" class="anchor"></a>Summarizing and determining the structure of a data frame.</h2>
<p>A <strong>data frame is the standard way in R to store tabular data</strong>. A data fame could also be thought of as a collection of vectors, all of which have the same length. Using only two functions, we can learn a lot about out data frame including some summary statistics as well as well as the “structure” of the data frame. Let’s examine what each of these functions can tell us:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## get summary statistics on a data frame</span>

<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></code></pre></div>
<pre><code>##   sample_id            CHROM                POS             ID         
##  Length:801         Length:801         Min.   :   1521   Mode:logical  
##  Class :character   Class :character   1st Qu.:1115970   NA's:801      
##  Mode  :character   Mode  :character   Median :2290361                 
##                                        Mean   :2243682                 
##                                        3rd Qu.:3317082                 
##                                        Max.   :4629225                 
##                                                                        
##      REF                ALT                 QUAL          FILTER       
##  Length:801         Length:801         Min.   :  4.385   Mode:logical  
##  Class :character   Class :character   1st Qu.:139.000   NA's:801      
##  Mode  :character   Mode  :character   Median :195.000                 
##                                        Mean   :172.276                 
##                                        3rd Qu.:225.000                 
##                                        Max.   :228.000                 
##                                                                        
##    INDEL              IDV              IMF               DP       
##  Mode :logical   Min.   : 2.000   Min.   :0.5714   Min.   : 2.00  
##  FALSE:700       1st Qu.: 7.000   1st Qu.:0.8824   1st Qu.: 7.00  
##  TRUE :101       Median : 9.000   Median :1.0000   Median :10.00  
##                  Mean   : 9.396   Mean   :0.9219   Mean   :10.57  
##                  3rd Qu.:11.000   3rd Qu.:1.0000   3rd Qu.:13.00  
##                  Max.   :20.000   Max.   :1.0000   Max.   :79.00  
##                  NA's   :700      NA's   :700                     
##       VDB                 RPB              MQB              BQB        
##  Min.   :0.0005387   Min.   :0.0000   Min.   :0.0000   Min.   :0.1153  
##  1st Qu.:0.2180410   1st Qu.:0.3776   1st Qu.:0.1070   1st Qu.:0.6963  
##  Median :0.4827410   Median :0.8663   Median :0.2872   Median :0.8615  
##  Mean   :0.4926291   Mean   :0.6970   Mean   :0.5330   Mean   :0.7784  
##  3rd Qu.:0.7598940   3rd Qu.:1.0000   3rd Qu.:1.0000   3rd Qu.:1.0000  
##  Max.   :0.9997130   Max.   :1.0000   Max.   :1.0000   Max.   :1.0000  
##                      NA's   :773      NA's   :773      NA's   :773     
##       MQSB              SGB               MQ0F           ICB         
##  Min.   :0.01348   Min.   :-0.6931   Min.   :0.00000   Mode:logical  
##  1st Qu.:0.95494   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:801      
##  Median :1.00000   Median :-0.6620   Median :0.00000                 
##  Mean   :0.96428   Mean   :-0.6444   Mean   :0.01127                 
##  3rd Qu.:1.00000   3rd Qu.:-0.6364   3rd Qu.:0.00000                 
##  Max.   :1.01283   Max.   :-0.4536   Max.   :0.66667                 
##  NA's   :48                                                          
##    HOB                AC          AN        DP4                  MQ       
##  Mode:logical   Min.   :1   Min.   :1   Length:801         Min.   :10.00  
##  NA's:801       1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00  
##                 Median :1   Median :1   Mode  :character   Median :60.00  
##                 Mean   :1   Mean   :1                      Mean   :58.19  
##                 3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00  
##                 Max.   :1   Max.   :1                      Max.   :60.00  
##                                                                           
##     Indiv              gt_PL               gt_GT   gt_GT_alleles     
##  Length:801         Length:801         Min.   :1   Length:801        
##  Class :character   Class :character   1st Qu.:1   Class :character  
##  Mode  :character   Mode  :character   Median :1   Mode  :character  
##                                        Mean   :1                     
##                                        3rd Qu.:1                     
##                                        Max.   :1                     
## </code></pre>
<p>Our data frame had 29 variables, so we get 29 fields that summarize the data. The <code>QUAL</code>, <code>IMF</code>, and <code>VDB</code> variables (and several others) are numerical data and so you get summary statistics on the min and max values for these columns, as well as mean, median, and interquartile ranges. Many of the other variables (e.g. <code>sample_id</code>) are treated as categorical data (which have special treatment in R - more on this in a bit). The most frequent 6 different categories and the number of times they appear (e.g. the sample_id called ‘SRR2584863’ appeared 25 times) are displayed. There was only one value for <code>CHROM</code>, “CP000819.1” which appeared in all 801 observations.</p>
<p>Before we operate on the data, we also need to know a little more about the data frame structure to do that we use the <code><a href="https://rdrr.io/r/utils/str.html">str()</a></code> function:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## get the structure of a data frame</span>

<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></code></pre></div>
<pre><code>## 'data.frame':    801 obs. of  29 variables:
##  $ sample_id    : chr  "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" ...
##  $ CHROM        : chr  "CP000819.1" "CP000819.1" "CP000819.1" "CP000819.1" ...
##  $ POS          : int  9972 263235 281923 433359 473901 648692 1331794 1733343 2103887 2333538 ...
##  $ ID           : logi  NA NA NA NA NA NA ...
##  $ REF          : chr  "T" "G" "G" "CTTTTTTT" ...
##  $ ALT          : chr  "G" "T" "T" "CTTTTTTTT" ...
##  $ QUAL         : num  91 85 217 64 228 210 178 225 56 167 ...
##  $ FILTER       : logi  NA NA NA NA NA NA ...
##  $ INDEL        : logi  FALSE FALSE FALSE TRUE TRUE FALSE ...
##  $ IDV          : int  NA NA NA 12 9 NA NA NA 2 7 ...
##  $ IMF          : num  NA NA NA 1 0.9 ...
##  $ DP           : int  4 6 10 12 10 10 8 11 3 7 ...
##  $ VDB          : num  0.0257 0.0961 0.7741 0.4777 0.6595 ...
##  $ RPB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ MQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ BQB          : num  NA 1 NA NA NA NA NA NA NA NA ...
##  $ MQSB         : num  NA NA 0.975 1 0.916 ...
##  $ SGB          : num  -0.556 -0.591 -0.662 -0.676 -0.662 ...
##  $ MQ0F         : num  0 0.167 0 0 0 ...
##  $ ICB          : logi  NA NA NA NA NA NA ...
##  $ HOB          : logi  NA NA NA NA NA NA ...
##  $ AC           : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ AN           : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ DP4          : chr  "0,0,0,4" "0,1,0,5" "0,0,4,5" "0,1,3,8" ...
##  $ MQ           : int  60 33 60 60 60 60 60 60 60 60 ...
##  $ Indiv        : chr  "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam" ...
##  $ gt_PL        : chr  "121,0" "112,0" "247,0" "91,0" ...
##  $ gt_GT        : int  1 1 1 1 1 1 1 1 1 1 ...
##  $ gt_GT_alleles: chr  "G" "T" "T" "CTTTTTTTT" ...</code></pre>
<p>Ok, thats a lot up unpack! Some things to notice.</p>
<ul>
<li>the object type <code>data.frame</code> is displayed in the first row along with its dimensions, in this case 801 observations (rows) and 29 variables (columns)</li>
<li>Each variable (column) has a name (e.g. <code>sample_id</code>). This is followed by the object mode (e.g. factor, int, num, etc.). Notice that before each variable name there is a <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> - this will be important later.</li>
</ul>
</div>
<div id="introducing-factors" class="section level2">
<h2 class="hasAnchor">
<a href="#introducing-factors" class="anchor"></a>Introducing Factors</h2>
<p>Factors are the final major data structure we will introduce in our R genomics lessons. Factors can be thought of as vectors which are specialized for categorical data. Given R’s specialization for statistics, this make sense since categorial and continuous variables usually have different treatments. Sometimes you may want to have data treated as a factor, but in other cases, this may be undesirable.</p>
<p>Since some of the data in our data frame are factors, lets see how factors work. First, we’ll extract one of the columns of our data frame to a new object, so that we don’t end up modifying the <code>variants</code> object by mistake.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## extract the "REF" column to a new object</span>

<span class="va">REF</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">$</span><span class="va">REF</span></code></pre></div>
<p>Let’s look at the first few items in our factor using <code><a href="https://rdrr.io/r/utils/head.html">head()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">REF</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "T"        "G"        "G"        "CTTTTTTT" "CCGC"     "C"</code></pre>
<p>What we get back are the items in our factor, and also something called “Levels”. <strong>Levels are the different categories contained in a factor</strong>. By default, R will organize the levels in a factor in alphabetical order. So the first level in this factor is “A”.</p>
<p>Lets look at the contents of a factor in a slightly different way using <code><a href="https://rdrr.io/r/utils/str.html">str()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">REF</span><span class="op">)</span></code></pre></div>
<pre><code>##  chr [1:801] "T" "G" "G" "CTTTTTTT" "CCGC" "C" "C" "G" ...</code></pre>
<p>For the sake of efficiency, R stores the content of a factor as a vector of integers, which an integer is assigned to each of the possible levels. Recall levels are assigned in alphabetical order. In this case, the first item in our “REF” object is “T”, which happens to be the 49th level of our factor, ordered alphabeticaly. The next two items are both “G”s, which is the 33rd level of our factor.</p>
</div>
<div id="plotting-and-ordering-factors" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-and-ordering-factors" class="anchor"></a>Plotting and ordering factors</h2>
<p>One of the most common uses for factors will be when you plot categorical values. For example, suppose we want to know how many of our variants had each possible nucleotide (or nucleotide combination) in the reference genome? We could generate a plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#plot(REF)</span></code></pre></div>
<p>This isn’t a particularly pretty example of a plot. We’ll be learning much more about creating nice, publication-quality graphics later in this lesson.</p>
<!-- For now, let's explore how we can order -->
<!-- the factors in our plot so that the first four values are "A", "C", "G", "T", with multi-nucleotide -->
<!-- combinations listed alphabetically after these four. -->
<!-- We can take our existing `REF` factor, and use the `factor()` -->
<!-- function again. This time we will pass it two new arguments: `levels` will be -->
<!-- assigned to a vector that has the REF values in the order we want them, -->
<!-- and we will set the `ordered` argument to TRUE. -->
<!-- ```{r, purl = FALSE} -->
<!--  # order the 'REF' factor to match our desired set of levels -->
<!-- REF <-  -->
<!-- ```         -->
<!-- We can now see the new ordering: -->
<!-- ```{r, purl = FALSE} -->
<!-- ``` -->
<!-- Although not all levels are shown, notice there are `<` signs indicating an -->
<!-- order. -->
</div>
<div id="subsetting-data-frames" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-data-frames" class="anchor"></a>Subsetting data frames</h2>
<p>Next, we are going to talk about how you can get specific values from data frames, and where necessary, change the mode of a column of values.</p>
<p>The first thing to remember is that a data frame is two-dimensional (rows and columns). Therefore, to select a specific value we will will once again use <code>[]</code> (bracket) notation, but we will specify more than one value (except in some cases where we are taking a range).</p>
<blockquote>
<h2 id="exercise-subsetting-a-data-frame">Exercise: Subsetting a data frame</h2>
<p><strong>Try the following indices and functions and try to figure out what they return</strong></p>
<ol style="list-style-type: lower-alpha">
<li><p><code>variants[1,1]</code></p></li>
<li><p><code>variants[2,4]</code></p></li>
<li><p><code>variants[801,29]</code></p></li>
<li><p><code>variants[2, ]</code></p></li>
<li><p><code>variants[-1, ]</code></p></li>
<li><p><code>variants[1:4,1]</code></p></li>
<li><p><code>variants[1:10,c("REF","ALT")]</code></p></li>
<li><p><code>variants[,c("sample_id")]</code></p></li>
<li><p><code><a href="https://rdrr.io/r/utils/head.html">head(variants)</a></code></p></li>
<li><p><code><a href="https://rdrr.io/r/utils/head.html">tail(variants)</a></code></p></li>
<li><p><code>variants$sample_id</code></p></li>
<li><p><code>variants[variants$REF == "A",]</code></p></li>
</ol>
</blockquote>
<p>The subsetting notation is very similar to what we learned for vectors. The key differences include:</p>
<ul>
<li>Typically provide two values separated by commas: data.frame[row, column]</li>
<li>In cases where you are taking a continuous range of numbers use a colon between the numbers (start:stop, inclusive)</li>
<li>For a non continuous set of numbers, pass a vector using <code><a href="https://rdrr.io/r/base/c.html">c()</a></code>
</li>
<li>Index using the name of a column(s) by passing them as vectors using <code><a href="https://rdrr.io/r/base/c.html">c()</a></code>
</li>
</ul>
<p>Finally, in all of the subsetting exercises above, we printed values to the screen. You can create a new data frame object by assigning them to a new object name:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create a new data frame containing only observations from SRR2584863 </span>

<span class="va">SRR2584863_variants</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>,<span class="op">]</span>

<span class="co"># check the dimension of the data frame</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 25 29</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get a summary of the data frame</span>

<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></code></pre></div>
<pre><code>##   sample_id            CHROM                POS             ID         
##  Length:25          Length:25          Min.   :   9972   Mode:logical  
##  Class :character   Class :character   1st Qu.:1331794   NA's:25       
##  Mode  :character   Mode  :character   Median :2618472                 
##                                        Mean   :2464989                 
##                                        3rd Qu.:3488669                 
##                                        Max.   :4616538                 
##                                                                        
##      REF                ALT                 QUAL         FILTER       
##  Length:25          Length:25          Min.   : 31.89   Mode:logical  
##  Class :character   Class :character   1st Qu.:104.00   NA's:25       
##  Mode  :character   Mode  :character   Median :211.00                 
##                                        Mean   :172.97                 
##                                        3rd Qu.:225.00                 
##                                        Max.   :228.00                 
##                                                                       
##    INDEL              IDV             IMF               DP      
##  Mode :logical   Min.   : 2.00   Min.   :0.6667   Min.   : 2.0  
##  FALSE:19        1st Qu.: 3.25   1st Qu.:0.9250   1st Qu.: 9.0  
##  TRUE :6         Median : 8.00   Median :1.0000   Median :10.0  
##                  Mean   : 7.00   Mean   :0.9278   Mean   :10.4  
##                  3rd Qu.: 9.75   3rd Qu.:1.0000   3rd Qu.:12.0  
##                  Max.   :12.00   Max.   :1.0000   Max.   :20.0  
##                  NA's   :19      NA's   :19                     
##       VDB               RPB              MQB               BQB        
##  Min.   :0.01627   Min.   :0.9008   Min.   :0.04979   Min.   :0.7507  
##  1st Qu.:0.07140   1st Qu.:0.9275   1st Qu.:0.09996   1st Qu.:0.7627  
##  Median :0.37674   Median :0.9542   Median :0.15013   Median :0.7748  
##  Mean   :0.40429   Mean   :0.9517   Mean   :0.39997   Mean   :0.8418  
##  3rd Qu.:0.65951   3rd Qu.:0.9771   3rd Qu.:0.57507   3rd Qu.:0.8874  
##  Max.   :0.99604   Max.   :1.0000   Max.   :1.00000   Max.   :1.0000  
##                    NA's   :22       NA's   :22        NA's   :22      
##       MQSB             SGB               MQ0F           ICB         
##  Min.   :0.5000   Min.   :-0.6904   Min.   :0.00000   Mode:logical  
##  1st Qu.:0.9599   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:25       
##  Median :0.9962   Median :-0.6620   Median :0.00000                 
##  Mean   :0.9442   Mean   :-0.6341   Mean   :0.04667                 
##  3rd Qu.:1.0000   3rd Qu.:-0.6168   3rd Qu.:0.00000                 
##  Max.   :1.0128   Max.   :-0.4536   Max.   :0.66667                 
##  NA's   :3                                                          
##    HOB                AC          AN        DP4                  MQ       
##  Mode:logical   Min.   :1   Min.   :1   Length:25          Min.   :10.00  
##  NA's:25        1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00  
##                 Median :1   Median :1   Mode  :character   Median :60.00  
##                 Mean   :1   Mean   :1                      Mean   :55.52  
##                 3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00  
##                 Max.   :1   Max.   :1                      Max.   :60.00  
##                                                                           
##     Indiv              gt_PL               gt_GT   gt_GT_alleles     
##  Length:25          Length:25          Min.   :1   Length:25         
##  Class :character   Class :character   1st Qu.:1   Class :character  
##  Mode  :character   Mode  :character   Median :1   Mode  :character  
##                                        Mean   :1                     
##                                        3rd Qu.:1                     
##                                        Max.   :1                     
## </code></pre>
</div>
<div id="coercing-values-in-data-frames" class="section level2">
<h2 class="hasAnchor">
<a href="#coercing-values-in-data-frames" class="anchor"></a>Coercing values in data frames</h2>
<blockquote>
<h2 id="tip-coercion-isnt-limited-to-data-frames">Tip: coercion isn’t limited to data frames</h2>
<p>While we are going to address coercion in the context of data frames most of these methods apply to other data structures, such as vectors</p>
</blockquote>
<p>Sometimes, it is possible that R will misinterpret the type of data represented in a data frame, or store that data in a mode which prevents you from operating on the data the way you wish. For example, a long list of gene names isn’t usually thought of as a categorical variable, the way that your experimental condition (e.g. control, treatment) might be. More importantly, some R packages you use to analyze your data may expect characters as input, not factors. At other times (such as plotting or some statistical analyses) a factor may be more appropriate. Ultimately, you should know how to change the mode of an object.</p>
<p>First, its very important to recognize that coercion happens in R all the time. This can be a good thing when R gets it right, or a bad thing when the result is not what you expect. Consider:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_chromosomes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'3'</span>, <span class="st">'11'</span>, <span class="st">'X'</span>, <span class="st">'6'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">snp_chromosomes</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
<p>Although there are several numbers in our vector, they are all in quotes, so we have explicitly told R to consider them as characters. However, even if we removed the quotes from the numbers, R would coerce everything into a character:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_chromosomes_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">11</span>, <span class="st">'X'</span>, <span class="fl">6</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">snp_chromosomes_2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_chromosomes_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "3"</code></pre>
<p>We can use the <code>as.</code> functions to explicitly coerce values from one form into another. Consider the following vector of characters, which all happen to be valid numbers:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"8762685"</span>, <span class="st">"66560624"</span>, <span class="st">"67545785"</span>, <span class="st">"154039662"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "8762685"</code></pre>
<p>Now we can coerce <code>snp_positions_2</code> into a numeric type using <code><a href="https://rdrr.io/r/base/numeric.html">as.numeric()</a></code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "double"</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_positions_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] 8762685</code></pre>
<p>Sometimes coercion is straight forward, but what would happen if we tried using <code><a href="https://rdrr.io/r/base/numeric.html">as.numeric()</a></code> on <code>snp_chromosomes_2</code></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_chromosomes_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">snp_chromosomes_2</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: NAs durch Umwandlung erzeugt</code></pre>
<p>If we check, we will see that an <code>NA</code> value (R’s default value for missing data) has been introduced.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp_chromosomes_2</span></code></pre></div>
<pre><code>## [1]  3 11 NA  6</code></pre>
<p>Trouble can really start when we try to coerce a factor. For example, when we try to coerce the <code>sample_id</code> column in our data frame into a numeric mode look at the result:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: NAs durch Umwandlung erzeugt</code></pre>
<pre><code>##   [1] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
##  [26] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
##  [51] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
##  [76] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [101] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [126] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [151] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [176] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [201] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [226] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [251] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [276] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [301] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [326] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [351] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [376] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [401] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [426] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [451] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [476] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [501] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [526] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [551] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [576] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [601] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [626] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [651] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [676] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [701] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [726] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [751] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [776] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
## [801] NA</code></pre>
<p>Strangely, it works! Almost. Instead of giving an error message, R returns numeric values, which in this case are the integers assigned to the levels in this factor. This kind of behavior can lead to hard-to-find bugs, for example when we do have numbers in a factor, and we get numbers from a coercion. If we don’t look carefully, we may not notice a problem.</p>
<p>If you need to coerce an entire column you can overwrite it using an expression like this one:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># make the 'REF' column a character type column</span>
 
<span class="va">variants</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span><span class="op">)</span>

<span class="co"># check the type of the column</span>
<span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "character"</code></pre>
</div>
<div id="stringsasfactors-false" class="section level2">
<h2 class="hasAnchor">
<a href="#stringsasfactors-false" class="anchor"></a>StringsAsFactors = FALSE</h2>
<p>Lets summarize this section on coercion with a few take home messages.</p>
<ul>
<li>When you explicitly coerce one data type into another (this is known as <strong>explicit coercion</strong>), be careful to check the result. Ideally, you should try to see if its possible to avoid steps in your analysis that force you to coerce.<br>
</li>
<li>R will sometimes coerce without you asking for it. This is called (appropriately) <strong>implicit coercion</strong>. For example when we tried to create a vector with multiple data types, R chose one type through implicit coercion.</li>
<li>Check the structure (<code><a href="https://rdrr.io/r/utils/str.html">str()</a></code>) of your data frames before working with them!</li>
</ul>
<p>Regarding the first bullet point, one way to avoid needless coercion when importing a data frame using any one of the <code><a href="https://rdrr.io/r/utils/read.table.html">read.table()</a></code> functions such as <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> is to set the argument <code>StringsAsFactors</code> to FALSE. By default, this argument is TRUE. Setting it to FALSE will treat any non-numeric column to a character type. <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> documentation, you will also see you can explicitly type your columns using the <code>colClasses</code> argument. Other R packages (such as the Tidyverse “readr”) don’t have this particular conversion issue, but many packages will still try to guess a data type.</p>
</div>
<div id="data-frame-bonus-material-math-sorting-renaming" class="section level2">
<h2 class="hasAnchor">
<a href="#data-frame-bonus-material-math-sorting-renaming" class="anchor"></a>Data frame bonus material: math, sorting, renaming</h2>
<p>Here are a few operations that don’t need much explanation, but which are good to know.</p>
<p>There are lots of arithmetic functions you may want to apply to your data frame, covering those would be a course in itself (there is some starting material <a href="https://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth/">here</a>). Our lessons will cover some additional summary statistical functions in a subsequent lesson, but overall we will focus on data cleaning and visualization.</p>
<p>You can use functions like <code><a href="https://rdrr.io/r/base/mean.html">mean()</a></code>, <code><a href="https://rdrr.io/r/base/Extremes.html">min()</a></code>, <code><a href="https://rdrr.io/r/base/Extremes.html">max()</a></code> on an individual column. Let’s look at the “DP” or filtered depth. This value shows the number of filtered reads that support each of the reported variants.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 79</code></pre>
<p>You can sort a data frame using the <code><a href="https://rdrr.io/r/base/order.html">order()</a></code> function:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sorted_by_DP</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sorted_by_DP</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 2 2 2 2 2 2</code></pre>
<blockquote>
<h2 id="exercise">Exercise</h2>
<p>The <code><a href="https://rdrr.io/r/base/order.html">order()</a></code> function lists values in increasing order by default. Look at the documentation for this function and change <code>sorted_by_DP</code> to start with variants with the greatest filtered depth (“DP”).</p>
</blockquote>
<p>You can rename columns:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span> <span class="op">==</span> <span class="st">"sample_id"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"strain"</span>

<span class="co"># check the column name (hint names are returned as a vector)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "strain"        "CHROM"         "POS"           "ID"           
##  [5] "REF"           "ALT"           "QUAL"          "FILTER"       
##  [9] "INDEL"         "IDV"           "IMF"           "DP"           
## [13] "VDB"           "RPB"           "MQB"           "BQB"          
## [17] "MQSB"          "SGB"           "MQ0F"          "ICB"          
## [21] "HOB"           "AC"            "AN"            "DP4"          
## [25] "MQ"            "Indiv"         "gt_PL"         "gt_GT"        
## [29] "gt_GT_alleles"</code></pre>
</div>
<div id="saving-your-data-frame-to-a-file" class="section level2">
<h2 class="hasAnchor">
<a href="#saving-your-data-frame-to-a-file" class="anchor"></a>Saving your data frame to a file</h2>
<p>We can save data to a file. We will save our <code>SRR2584863_variants</code> object to a .csv file using the <code><a href="https://rdrr.io/r/utils/write.table.html">write.csv()</a></code> function:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span><span class="op">(</span><span class="va">SRR2584863_variants</span>, file <span class="op">=</span> <span class="st">"SRR2584863_variants.csv"</span><span class="op">)</span></code></pre></div>
<p>The <code><a href="https://rdrr.io/r/utils/write.table.html">write.csv()</a></code> function has some additional arguments listed in the help, but at a minimum you need to tell it what data frame to write to file, and give a path to a file name in quotes (if you only provide a file name, the file will be written in the current working directory).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Klaus Schliep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
