<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Aggregating and Analyzing Data with dplyr • Metagenomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Aggregating and Analyzing Data with dplyr">
<meta property="og:description" content="Metagenomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Metagenomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01-introduction.html">Introducing R and RStudio</a>
    </li>
    <li>
      <a href="../articles/02-r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="../articles/03-basics-factors-dataframes.html">R Basics continued - factors and data frames</a>
    </li>
    <li>
      <a href="../articles/04-dplyr.html">Aggregating and Analyzing Data with dplyr</a>
    </li>
    <li>
      <a href="../articles/05-data-visualization.html">Data Visualization with ggplot2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="04-dplyr_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Aggregating and Analyzing Data with dplyr</h1>
                        <h4 class="author">Klaus Schliep</h4>
            
            <h4 class="date">Mai 17, 2021</h4>
      
      
      <div class="hidden name"><code>04-dplyr.Rmd</code></div>

    </div>

    
    
<p>Bracket subsetting is handy, but it can be cumbersome and difficult to read, especially for complicated operations.</p>
<p>Luckily, the <a href="https://cran.r-project.org/package=dplyr"><code>dplyr</code></a> package provides a number of very useful functions for manipulating dataframes in a way that will reduce repetition, reduce the probability of making errors, and probably even save you some typing. As an added bonus, you might even find the <code>dplyr</code> grammar easier to read.</p>
<p>Here we’re going to cover 6 of the most commonly used functions as well as using pipes (<code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code>) to combine them.</p>
<ol style="list-style-type: decimal">
<li><code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code></li>
<li><code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code></li>
</ol>
<p>Packages in R are sets of additional functions that let you do more stuff in R. The functions we’ve been using, like <code><a href="https://rdrr.io/r/utils/str.html">str()</a></code>, come built into R; packages give you access to more functions. You need to install a package and then load it to be able to use it.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span> <span class="co">## install</span></code></pre></div>
<p>You might get asked to choose a CRAN mirror – this is asking you to choose a site to download the package from. The choice doesn’t matter too much; I’d recommend choosing the RStudio mirror.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org">"dplyr"</a></span><span class="op">)</span>          <span class="co">## load</span></code></pre></div>
<p>You only need to install a package once per computer, but you need to load it every time you open a new R session and want to use that package.</p>
<div id="what-is-dplyr" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-dplyr" class="anchor"></a>What is dplyr?</h2>
<p>The package <code>dplyr</code> is a fairly new (2014) package that tries to provide easy tools for the most common data manipulation tasks. It is built to work directly with data frames. The thinking behind it was largely inspired by the package <code>plyr</code> which has been in use for some time but suffered from being slow in some cases.<code>dplyr</code> addresses this by porting much of the computation to C++. An additional feature is the ability to work with data stored directly in an external database. The benefits of doing this are that the data can be managed natively in a relational database, queries can be conducted on that database, and only the results of the query returned.</p>
<p>This addresses a common problem with R in that all operations are conducted in memory and thus the amount of data you can work with is limited by available memory. The database connections essentially remove that limitation in that you can have a database of many 100s GB, conduct queries on it directly and pull back just what you need for analysis in R.</p>
<div id="selecting-columns-and-filtering-rows" class="section level3">
<h3 class="hasAnchor">
<a href="#selecting-columns-and-filtering-rows" class="anchor"></a>Selecting columns and filtering rows</h3>
<p>To select columns of a data frame, use <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code>. The first argument to this function is the data frame (<code>variants</code>), and the subsequent arguments are the columns to keep.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></code></pre></div>
<pre><code>##    sample_id      REF       ALT DP
## 1 SRR2584863        T         G  4
## 2 SRR2584863        G         T  6
## 3 SRR2584863        G         T 10
## 4 SRR2584863 CTTTTTTT CTTTTTTTT 12
## 5 SRR2584863     CCGC    CCGCGC 10
## 6 SRR2584863        C         T 10</code></pre>
<p>To select all columns <em>except</em> certain ones, put a “-” in front of the variable to exclude it.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">variants</span>, <span class="op">-</span><span class="va">CHROM</span><span class="op">)</span></code></pre></div>
<pre><code>##    sample_id    POS ID      REF       ALT QUAL FILTER INDEL IDV IMF DP
## 1 SRR2584863   9972 NA        T         G   91     NA FALSE  NA  NA  4
## 2 SRR2584863 263235 NA        G         T   85     NA FALSE  NA  NA  6
## 3 SRR2584863 281923 NA        G         T  217     NA FALSE  NA  NA 10
## 4 SRR2584863 433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0 12
## 5 SRR2584863 473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9 10
## 6 SRR2584863 648692 NA        C         T  210     NA FALSE  NA  NA 10
##         VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
## 1 0.0257451  NA  NA  NA       NA -0.556411 0.000000  NA  NA  1  1 0,0,0,4 60
## 2 0.0961330   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
## 3 0.7740830  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
## 4 0.4777040  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
## 5 0.6595050  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
## 6 0.2680140  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
##                                                                Indiv gt_PL
## 1 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 121,0
## 2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
## 3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
## 4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
## 5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
## 6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
##   gt_GT gt_GT_alleles
## 1     1             G
## 2     1             T
## 3     1             T
## 4     1     CTTTTTTTT
## 5     1        CCGCGC
## 6     1             T</code></pre>
<p>To choose rows, use <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span></code></pre></div>
<pre><code>##    sample_id      CHROM    POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
## 1 SRR2584863 CP000819.1   9972 NA        T         G   91     NA FALSE  NA  NA
## 2 SRR2584863 CP000819.1 263235 NA        G         T   85     NA FALSE  NA  NA
## 3 SRR2584863 CP000819.1 281923 NA        G         T  217     NA FALSE  NA  NA
## 4 SRR2584863 CP000819.1 433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
## 5 SRR2584863 CP000819.1 473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
## 6 SRR2584863 CP000819.1 648692 NA        C         T  210     NA FALSE  NA  NA
##   DP       VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
## 1  4 0.0257451  NA  NA  NA       NA -0.556411 0.000000  NA  NA  1  1 0,0,0,4 60
## 2  6 0.0961330   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
## 3 10 0.7740830  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
## 4 12 0.4777040  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
## 5 10 0.6595050  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
## 6 10 0.2680140  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
##                                                                Indiv gt_PL
## 1 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 121,0
## 2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
## 3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
## 4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
## 5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
## 6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
##   gt_GT gt_GT_alleles
## 1     1             G
## 2     1             T
## 3     1             T
## 4     1     CTTTTTTTT
## 5     1        CCGCGC
## 6     1             T</code></pre>
<p>Note that this is equivalent to the base R code below, but is easier to read!</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span><span class="op">[</span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>,<span class="op">]</span></code></pre></div>
</div>
<div id="pipes" class="section level3">
<h3 class="hasAnchor">
<a href="#pipes" class="anchor"></a>Pipes</h3>
<p>But what if you wanted to select and filter? We can do this with pipes. Pipes, are a fairly recent addition to R. Pipes let you take the output of one function and send it directly to the next, which is useful when you need to many things to the same data set. It was possible to do this before pipes were added to R, but it was much messier and more difficult. Pipes in R look like <code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code> and are made available via the <code>magrittr</code> package, which is installed as part of <code>dplyr</code>. If you use RStudio, you can type the pipe with <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you’re using a PC, or <kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you’re using a Mac.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## Rows: 25
## Columns: 3
## $ REF &lt;chr&gt; "T", "G", "G", "CTTTTTTT", "CCGC", "C", "C", "G", "ACAGCCAGCCAGCCA…
## $ ALT &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "ACAGCCAGCCAG…
## $ DP  &lt;int&gt; 4, 6, 10, 12, 10, 10, 8, 11, 3, 7, 9, 20, 12, 19, 15, 10, 14, 9, 1…</code></pre>
<p>In the above code, we use the pipe to send the <code>variants</code> dataset first through <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code>, to keep rows where <code>sample_id</code> matches a particular sample, and then through <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> to keep only the <code>REF</code>, <code>ALT</code>, and <code>DP</code> columns. Since <code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code> takes the object on its left and passes it as the first argument to the function on its right, we don’t need to explicitly include the data frame as an argument to the <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> functions any more. We then pipe the results to the <code><a href="https://rdrr.io/r/utils/head.html">head()</a></code> function so that we only see the first six rows of data.</p>
<p>Some may find it helpful to read the pipe like the word “then”. For instance, in the above example, we took the data frame <code>variants</code>, <em>then</em> we <code>filter</code>ed for rows where <code>sample_id</code> was SRR2584863, <em>then</em> we <code>select</code>ed the <code>REF</code>, <code>ALT</code>, and <code>DP</code> columns, <em>then</em> we showed only the first six rows. The <strong><code>dplyr</code></strong> functions by themselves are somewhat simple, but by combining them into linear workflows with the pipe, we can accomplish more complex manipulations of data frames.</p>
<p>If we want to create a new object with this smaller version of the data we can do so by assigning it a new name:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SRR2584863_variants</span> <span class="op">&lt;-</span> <span class="va">variants</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></code></pre></div>
<p>This new object includes all of the data from this sample. Let’s look at just the first six rows to confirm it’s what we want:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></code></pre></div>
<pre><code>##        REF       ALT DP
## 1        T         G  4
## 2        G         T  6
## 3        G         T 10
## 4 CTTTTTTT CTTTTTTTT 12
## 5     CCGC    CCGCGC 10
## 6        C         T 10</code></pre>
<blockquote>
<h2 id="exercise-pipe-and-filter">Exercise: Pipe and filter</h2>
<p>Starting with the <code>variants</code> dataframe, use pipes to subset the data to include only observations from SRR2584863 sample, where the filtered depth (DP) is at least 10. Retain only the columns <code>REF</code>, <code>ALT</code>, and <code>POS</code>.</p>
</blockquote>
</div>
<div id="mutate" class="section level3">
<h3 class="hasAnchor">
<a href="#mutate" class="anchor"></a>Mutate</h3>
<p>Frequently you’ll want to create new columns based on the values in existing columns, for example to do unit conversions or find the ratio of values in two columns. For this we’ll use the <code>dplyr</code> function <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code>.</p>
<p>We have a column titled “QUAL”. This is a Phred-scaled confidence score that a polymorphism exists at this position given the sequencing data. Lower QUAL scores indicate low probability of a polymorphism existing at that site. We can convert the confidence value QUAL to a probability value according to the formula:</p>
<p>Probability = 1- 10 ^ -(QUAL/10)</p>
<p>Let’s add a column (<code>POLPROB</code>) to our <code>variants</code> dataframe that shows the probability of a polymorphism at that site given the data. We’ll show only the first six rows of data.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>POLPROB <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="op">(</span><span class="fl">10</span> <span class="op">^</span> <span class="op">-</span><span class="op">(</span><span class="va">QUAL</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## Rows: 801
## Columns: 30
## $ sample_id     &lt;chr&gt; "SRR2584863", "SRR2584863", "SRR2584863", "SRR2584863", …
## $ CHROM         &lt;chr&gt; "CP000819.1", "CP000819.1", "CP000819.1", "CP000819.1", …
## $ POS           &lt;int&gt; 9972, 263235, 281923, 433359, 473901, 648692, 1331794, 1…
## $ ID            &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ REF           &lt;chr&gt; "T", "G", "G", "CTTTTTTT", "CCGC", "C", "C", "G", "ACAGC…
## $ ALT           &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…
## $ QUAL          &lt;dbl&gt; 91.0000, 85.0000, 217.0000, 64.0000, 228.0000, 210.0000,…
## $ FILTER        &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ INDEL         &lt;lgl&gt; FALSE, FALSE, FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, TR…
## $ IDV           &lt;int&gt; NA, NA, NA, 12, 9, NA, NA, NA, 2, 7, NA, NA, NA, NA, NA,…
## $ IMF           &lt;dbl&gt; NA, NA, NA, 1.000000, 0.900000, NA, NA, NA, 0.666667, 1.…
## $ DP            &lt;int&gt; 4, 6, 10, 12, 10, 10, 8, 11, 3, 7, 9, 20, 12, 19, 15, 10…
## $ VDB           &lt;dbl&gt; 0.0257451, 0.0961330, 0.7740830, 0.4777040, 0.6595050, 0…
## $ RPB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.900802, …
## $ MQB           &lt;dbl&gt; NA, 1.0000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.1501340…
## $ BQB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.750668, …
## $ MQSB          &lt;dbl&gt; NA, NA, 0.974597, 1.000000, 0.916482, 0.916482, 0.900802…
## $ SGB           &lt;dbl&gt; -0.556411, -0.590765, -0.662043, -0.676189, -0.662043, -…
## $ MQ0F          &lt;dbl&gt; 0.000000, 0.166667, 0.000000, 0.000000, 0.000000, 0.0000…
## $ ICB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ HOB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ AC            &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
## $ AN            &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
## $ DP4           &lt;chr&gt; "0,0,0,4", "0,1,0,5", "0,0,4,5", "0,1,3,8", "1,0,2,7", "…
## $ MQ            &lt;int&gt; 60, 33, 60, 60, 60, 60, 60, 60, 60, 60, 25, 60, 10, 60, …
## $ Indiv         &lt;chr&gt; "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned…
## $ gt_PL         &lt;chr&gt; "121,0", "112,0", "247,0", "91,0", "255,0", "240,0", "20…
## $ gt_GT         &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
## $ gt_GT_alleles &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…
## $ POLPROB       &lt;dbl&gt; 1.0000000, 1.0000000, 1.0000000, 0.9999996, 1.0000000, 1…</code></pre>
<blockquote>
<h2 id="exercise">Exercise</h2>
<p>There are a lot of columns in our dataset, so let’s just look at the <code>sample_id</code>, <code>POS</code>, <code>QUAL</code>, and <code>POLPROB</code> columns for now. Add a line to the above code to only show those columns.</p>
</blockquote>
</div>
<div id="split-apply-combine-data-analysis-and-the-summarize-function" class="section level3">
<h3 class="hasAnchor">
<a href="#split-apply-combine-data-analysis-and-the-summarize-function" class="anchor"></a>Split-apply-combine data analysis and the summarize() function</h3>
<p>Many data analysis tasks can be approached using the “split-apply-combine” paradigm: split the data into groups, apply some analysis to each group, and then combine the results. <code>dplyr</code> makes this very easy through the use of the <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> function, which splits the data into groups. When the data is grouped in this way <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code> can be used to collapse each group into a single-row summary. <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code> does this by applying an aggregating or summary function to each group. For example, if we wanted to group by sample_id and find the number of rows of data for each sample, we would do:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   sample_id  `n()`
##   &lt;chr&gt;      &lt;int&gt;
## 1 SRR2584863    25
## 2 SRR2584866   766
## 3 SRR2589044    10</code></pre>
<p>Here the summary function used was <code><a href="https://dplyr.tidyverse.org/reference/context.html">n()</a></code> to find the count for each group. We can also apply many other functions to individual columns to get other summary statistics. For example, we can use built-in functions like <code><a href="https://rdrr.io/r/base/mean.html">mean()</a></code>, <code><a href="https://rdrr.io/r/stats/median.html">median()</a></code>, <code><a href="https://rdrr.io/r/base/Extremes.html">min()</a></code>, and <code><a href="https://rdrr.io/r/base/Extremes.html">max()</a></code>. These are called “built-in functions” because they come with R and don’t require that you install any additional packages. By default, all <strong>R functions operating on vectors that contains missing data will return NA</strong>. It’s a way to make sure that users know they have missing data, and make a conscious decision on how to deal with it. When dealing with simple statistics like the mean, the easiest way to ignore <code>NA</code> (the missing data) is to use <code>na.rm = TRUE</code> (<code>rm</code> stands for remove).</p>
<p>So to view the highest filtered depth (<code>DP</code>) for each sample:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">DP</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   sample_id  `max(DP)`
##   &lt;chr&gt;          &lt;int&gt;
## 1 SRR2584863        20
## 2 SRR2584866        79
## 3 SRR2589044        16</code></pre>
<!-- You can group by multiple columns too. For example, we might want to  -->
<!-- know if ...  -->
<!-- ```{r, purl = FALSE} -->
<!-- ``` -->
<!-- You can also summarize multiple variables at the same time. Let's -->
<!-- count how many samples we have in each group using the `n()` function: -->
<!-- ```{r, purl = FALSE} -->
<!-- ``` -->
<p><a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf">Handy dplyr cheatsheet</a></p>
<p><em>Much of this lesson was copied or adapted from Jeff Hollister’s <a href="http://usepa.github.io/introR/2015/01/14/03-Clean/">materials</a></em></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Klaus Schliep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
